applications/FluidDynamicsApplication/custom_processes/shock_capturing_entropy_viscosity_process.h

```diff
@@ -0,0 +1,350 @@
+        }
+
+    private:
+        /**
+         * @brief Computes the divergence
+         *
+         * @param rShapeFunGradients The gradients of the shape functions [ndims x nnodes]
+         * @param rNodalValues Values of the magnitude at the nodes. [ndims x nnodes]
+         * @return Result
+         */
+        static double Divergence(const Matrix& rShapeFunGradients, const Matrix& rNodalValues);
+    };
+
+
+    /// Pointer definition of ShockCapturingEntropyViscosityProcess
+    KRATOS_CLASS_POINTER_DEFINITION(ShockCapturingEntropyViscosityProcess);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Constructor with model
+    ShockCapturingEntropyViscosityProcess(
+        Model& rModel,
+        Parameters rParameters)
+        : ShockCapturingEntropyViscosityProcess(rModel.GetModelPart(rParameters["model_part_name"].GetString()), rParameters) {};
+
+    /// Constructor with model part
+    ShockCapturingEntropyViscosityProcess(
+        ModelPart& rModelPart,
+        Parameters rParameters);
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    void ExecuteInitializeSolutionStep() override;
+
+    void ExecuteFinalizeSolutionStep() override;
+
+    int Check() override;
+
+    const Parameters GetDefaultParameters() const override;
+
+    ///@}
+    ///@name Access
+    ///@{
+
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "ShockCapturingEntropyViscosityProcess";
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override { rOStream << "ShockCapturingEntropyViscosityProcess"; }
+
+    /// Print object's data.
+    void PrintData(std::ostream&) const override {}
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+private:
+    ///@name Static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+    bool mFirstTimeStep = true;
+    bool mComputeAreasEveryStep = false;
+    double mEntropyConstant = 0.0;
+    double mEnergyConstant = 0.0;
+    double mArtificialMassDiffusivityPrandtl = 0.0;
+    double mArtificialConductivityPrandtl = 0.0;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    void ValidateAndAssignParameters(Parameters &rParameters);
+
+    void UpdateNodalAreaProcess();
+
+    /**
+     * @brief Computes nodal entropies and initializes artificial variables to 0.
+     *
+     * @tparam WriteBufferIndex: The buffer index to write the result in. This is an
+     * ugly hack to deal with the first time-step.
+     */
+    template<unsigned int WriteBufferIndex = 0>
```

Comment:
As mentioned above I'd avoid the template and pass the step as a function argument.